src/bpmncwpverify/visitors.py

```diff
@@ -99,25 +111,339 @@ def end_visit_bpmn(self, bpmn: Bpmn) -> None:
+                    guard += "&&".join(
+                        [
+                            "hasToken({})".format(self.get_location(element, loc))
+                            for loc in element.in_flows
+                        ]
+                    )
+                else:
+                    guard += "||".join(
+                        [
+                            "hasToken({})".format(self.get_location(element, loc))
+                            for loc in element.in_flows
+                        ]
+                    )
+                guard += " )\n"
+        if option_type != "Task":
+            for msg in element.in_msgs:
+                consume_locations.append(self.get_location(element, msg))
+            if element.in_msgs:
+                if element.in_flows or option_type == "Task-END":
+                    guard += "&&"
+                guard += "( "
+                guard += "&&".join(
+                    [
+                        "hasToken({})".format(self.get_location(element, loc))
+                        for loc in element.in_msgs
+                    ]
+                )
+                guard += " )\n"
+        if option_type == "XOR":
+            guard += "\t&&"
+            guard += "\t({}_hasOption)".format(element.name)
+        guard += ")"
+        if option_type != "Task-END":
+            behavior_inline = "{x}_BehaviorModel()".format(x=element.name)
+        if option_type == "Task":
+            put_flow_ids.append(element_id)
+            put_locations.append(self.get_location(element))
+            for msg in element.out_msgs:
+                put_flow_ids.append(msg.id)
+                put_locations.append(self.get_location(msg.target_node, msg))
+        else:
+            for flow in element.out_flows:
+                put_flow_ids.append(flow.id)
+                put_locations.append(self.get_location(flow.target_node, flow))
+                if option_type == "XOR":
+                    put_conditions.append(flow.name)
+            if option_type != "Task-END":
+                for msg in element.out_msgs:
+                    put_flow_ids.append(msg.id)
+                    put_locations.append(self.get_location(msg.target_node, msg))
+        if start_guard:
+            guard = start_guard
+            consume_locations.append(self.get_location(element))
+        self.write_workflow_lines(
+            self.create_option(
+                guard,
+                consume_locations,
+                behavior_inline,
+                put_conditions,
+                put_locations,
+                put_flow_ids,
+                element_id,
+                option_type,
+            )
+        )
+
+    def gen_behavior_model(self, place_name: str, update_state: bool = False) -> str:
+        ret = "//{}\n".format(place_name)
+        ret += "inline {x}_BehaviorModel(){{\n".format(x=place_name)
+        # TODO: figure out how to do this without modifies file
```

Comment:
I know now how we are going to inject the behavior model (mostly). I'm going to use the out-going edge from a task to assert the `post-condition` that must hold leaving the task. We can generate from that the Promela code. The only think we need to figure out is how we want to specify non-determinism. I'm thinking of something like x in {1, 2, 3} type notation saying that x has one of the values in the set.

Reply:
ah, yes, this makes sense